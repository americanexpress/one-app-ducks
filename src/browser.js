/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { fromJS } from 'immutable';
import Cookies from 'js-cookie';
import typeScope from './utils/typeScope';

// constants
const SET_ORIGIN = `${typeScope}/browser/SET_ORIGIN`;
const SET_USER_AGENT = `${typeScope}/browser/SET_USER_AGENT`;

function buildInitialState({ req } = {}) {
  let cookies = null;
  let userAgent = null;
  let location = {
    origin: null,
    host: null,
    protocol: null,
  };

  if (global.BROWSER) {
    cookies = Cookies.get();
    userAgent = window.navigator.userAgent;
    location = {
      origin: window.location.origin
        || `${window.location.protocol}//${window.location.host}`,
      host: window.location.host,
      protocol: window.location.protocol,
    };
  } else if (req) {
    cookies = req.cookies;
    userAgent = req.headers['user-agent'];
    const host = req.forwarded && req.forwarded.host ? req.forwarded.host : req.headers.host;
    const protocol = req.forwarded && req.forwarded.proto ? req.forwarded.proto : req.protocol;
    location = {
      origin: `${protocol}://${host}`,
      host,
      protocol: `${protocol}:`,
    };
  }

  return fromJS({
    cookies,
    location,
    userAgent,
  });
}

// reducers
// eslint-disable-next-line default-param-last -- reducers have default params first
export default function browserReducer(state = buildInitialState(), action) {
  switch (action.type) {
    case SET_ORIGIN:
      return state.setIn(['location', 'origin'], action.origin);

    case SET_USER_AGENT:
      return state.set('userAgent', action.userAgent);

    default:
      return fromJS(state);
  }
}

browserReducer.buildInitialState = buildInitialState;

// actions
export function setOrigin(origin) {
  return {
    type: SET_ORIGIN,
    origin,
  };
}

export function setUserAgent(userAgent) {
  return {
    type: SET_USER_AGENT,
    userAgent,
  };
}

// selectors
const getCookieState = (state) => state.getIn(['browser', 'cookies']).toJS();

export const getCookies = (state) => (typeof window !== 'undefined' ? Cookies.get() : getCookieState(state));
