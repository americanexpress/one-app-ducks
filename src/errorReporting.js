/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

/* eslint no-bitwise: ["error", { "int32Hint": true }] */
import { fromJS } from 'immutable';
import util from 'util';
import typeScope from './utils/typeScope';

// action constants
export const ADD_ERROR_REPORT_TO_QUEUE = `${typeScope}/error-reporting/ADD_ERROR_REPORT_TO_QUEUE`;
export const SCHEDULE_ERROR_REPORT = `${typeScope}/error-reporting/SCHEDULE_ERROR_REPORT`;

// CALL_API constants
export const SEND_ERROR_REPORT_REQUEST = `${typeScope}/error-reporting/SEND_ERROR_REPORT_REQUEST`;
export const SEND_ERROR_REPORT_SUCCESS = `${typeScope}/error-reporting/SEND_ERROR_REPORT_SUCCESS`;
export const SEND_ERROR_REPORT_FAILURE = `${typeScope}/error-reporting/SEND_ERROR_REPORT_FAILURE`;

const DEFAULT_REQUEST_WAIT = 500; // ms
const MAX_RETRY = 10e3;

const defaultState = fromJS({
  queue: [],
  pending: [],
  pendingPromise: null,
  retryWait: DEFAULT_REQUEST_WAIT,
});

export function formatErrorReport(error, otherData) {
  return {
    msg: error && error.message,
    // TODO: use StackTrace to format the stack?
    stack: error && error.stack, // IE >= 10
    href: global.BROWSER ? global.location.href : undefined,
    otherData,
  };
}

export default function errorReportingReducer(state = defaultState, action) {
  const stateQueue = state.get('queue');
  const statePending = state.get('pending');

  switch (action.type) {
    case ADD_ERROR_REPORT_TO_QUEUE:
      if (!(action.error || action.otherData)) {
        console.warn('no error xor otherData given to report, probably due to localhost quirks');
        return state;
      }
      return state.set(
        'queue',
        stateQueue.push(fromJS(formatErrorReport(action.error, action.otherData)))
      );

    case SEND_ERROR_REPORT_REQUEST:
      return state
        .set('pending', stateQueue)
        .set('pendingPromise', action.promise)
        .set('queue', fromJS([]));

    case SEND_ERROR_REPORT_SUCCESS:
      return state
        .set('retryWait', DEFAULT_REQUEST_WAIT)
        .set('pending', fromJS([]))
        .set('pendingPromise', null);

    case SEND_ERROR_REPORT_FAILURE:
      return state
        .set('pending', fromJS([]))
        .set('queue', stateQueue.concat(statePending))
        .set('pendingPromise', null);

    case SCHEDULE_ERROR_REPORT:
      return state
        .set('pendingPromise', action.promise)
        .set(
          'retryWait',
          Math.min(state.get('retryWait') * (1.25 + (Math.random() / 3)) | 0, MAX_RETRY)
        );

    default:
      return state;
  }
}

function getPendingPromise(state) {
  return state.getIn(['errorReporting', 'pendingPromise']);
}

export function serverSideError(error) {
  // nodejs console truncates output by default
  // https://nodejs.org/api/util.html#util_util_inspect_object_options
  console.error(util.inspect(error, false, 10, true));
  return Promise.resolve({ thankYou: true });
}

function thenSendErrorReport({
  promise, dispatch, getState, fetchClient,
}) {
  return promise
    .then(() => {
      const state = getState();
      const reportingUrl = state.getIn(['config', 'reportingUrl']);
      const queue = state.getIn(['errorReporting', 'queue']).toJS();

      if (queue.length === 0) {
        return null;
      }

      let requestPromise;
      if (global.BROWSER) {
        requestPromise = fetchClient(reportingUrl, {
          method: 'post',
          body: JSON.stringify(queue),
          headers: {
            'Content-Type': 'application/json',
          },
        }).then((response) => {
          if (response.ok) {
            return response.text();
          }
          return Promise.reject(response);
        });
      } else {
        requestPromise = serverSideError(queue);
      }

      const delayPromise = new Promise((resolve) => setTimeout(resolve, DEFAULT_REQUEST_WAIT));
      dispatch({
        type: SEND_ERROR_REPORT_REQUEST,
        payload: JSON.stringify(queue),
        promise: Promise.all([requestPromise, delayPromise]),
      });

      return requestPromise;
    })
    .then((body) => {
      dispatch({
        type: SEND_ERROR_REPORT_SUCCESS,
        data: body,
      });
      return body;
    })
    .catch(() => {
    // catch the error to prevent looping
    // and attempt a retry
      const retryPromise = new Promise((res) => {
        setTimeout(() => res(), getState().getIn(['errorReporting', 'retryWait']));
      });

      dispatch({
        type: SCHEDULE_ERROR_REPORT,
        promise: thenSendErrorReport({
          promise: retryPromise, dispatch, getState, fetchClient,
        }),
      });
      return retryPromise;
    });
}

export function sendErrorReport() {
  // FIXME: figure out how to debounce (min waiting time between processing queues)
  return function getSendErrorReportPromise(dispatch, getState, { fetchClient }) {
    const pendingPromise = getPendingPromise(getState());

    if (pendingPromise) {
      // there's already a request out
      return thenSendErrorReport({
        promise: pendingPromise, dispatch, getState, fetchClient,
      });
    }

    return thenSendErrorReport({
      promise: Promise.resolve(), dispatch, getState, fetchClient,
    });
  };
}

export function addErrorToReport(error, otherData) {
  return function addToQueueAndScheduleReport(dispatch) {
    dispatch({
      type: ADD_ERROR_REPORT_TO_QUEUE,
      error,
      otherData,
    });

    return dispatch(sendErrorReport());
  };
}
