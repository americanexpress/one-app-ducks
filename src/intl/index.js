/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { fromJS, Map as iMap } from 'immutable';

import typeScope from '../utils/typeScope';
import { getModuleBaseUrl } from '../utils/modules';

export const defaultLocale = 'en-US';

/* istanbul ignore next */
// This import will resolve in the build output
// eslint-disable-next-line import/no-unresolved -- load based on server or browser
const localePacks = global.BROWSER ? require('./localePacks.client') : require('./localePacks.node');

export const UPDATE_LOCALE = `${typeScope}/intl/UPDATE_LOCALE`;
export const LANGUAGE_PACK_REQUEST = `${typeScope}/intl/LANGUAGE_PACK_REQUEST`;
export const LANGUAGE_PACK_SUCCESS = `${typeScope}/intl/LANGUAGE_PACK_SUCCESS`;
export const LANGUAGE_PACK_FAILURE = `${typeScope}/intl/LANGUAGE_PACK_FAILURE`;
export const LANGUAGE_PACK_DEFERRED_FORCE_LOAD = `${typeScope}/intl/LANGUAGE_PACK_DEFERRED_FORCE_LOAD`;

function buildInitialState({ req } = {}) {
  let activeLocale = null;

  if (!req) {
    const { navigator } = global;
    activeLocale = (navigator && navigator.language) || defaultLocale;
  } else {
    const acceptsLanguages = req.acceptsLanguages();
    activeLocale = acceptsLanguages && acceptsLanguages[0];
    if (!activeLocale || activeLocale === '*') {
      activeLocale = defaultLocale;
    }
  }

  return fromJS({
    activeLocale,
  });
}
// eslint-disable-next-line default-param-last -- reducers have default params first
export default function reducer(state = buildInitialState(), action) {
  switch (action.type) {
    case UPDATE_LOCALE: {
      return state.set('activeLocale', action.locale);
    }

    case LANGUAGE_PACK_REQUEST: {
      const { locale, componentKey, promise } = action;
      const langPackState = state.getIn(['languagePacks', locale, componentKey], iMap());
      return langPackState.get('_loadedOnServer')
        ? state
        : state.updateIn(['languagePacks', locale, componentKey], iMap(), (nextState) => nextState.withMutations((map) => map
          .set('data', map.get('data', iMap()))
          .set('isLoading', true)
          .set('promise', promise)
          .delete('error')
          .delete('errorExpiration')
        ));
    }

    case LANGUAGE_PACK_SUCCESS: {
      const { locale, componentKey, data } = action;
      return state.updateIn(['languagePacks', locale, componentKey], iMap(), (nextState) => nextState.withMutations((map) => map
        .set('data', iMap(data))
        .set('isLoading', false)
        .set('_loadedOnServer', !global.BROWSER)
        .delete('promise')
        .delete('error')
        .delete('errorExpiration')
      ));
    }

    case LANGUAGE_PACK_FAILURE: {
      const { locale, componentKey, error } = action;

      if (state.getIn(['languagePacks', locale, componentKey, '_loadedOnServer'])) {
        return state;
      }
      return state.updateIn(['languagePacks', locale, componentKey], iMap(), (nextState) => nextState.withMutations((map) => map
        .set('data', map.get('data', iMap()))
        .set('isLoading', false)
        .set('errorExpiration', Date.now() + 10e3)
        .set('error', error)
        .delete('promise')
      ));
    }

    case LANGUAGE_PACK_DEFERRED_FORCE_LOAD: {
      const { locale, componentKey } = action;
      return state.setIn(['languagePacks', locale, componentKey, '_pendingDeferredForceLoad'], true);
    }

    default:
      return state;
  }
}

reducer.buildInitialState = buildInitialState;

// when running on the server, store previous requests for a language pack
// reduces network errors causing a needless 500 and removes the I/O time
/* istanbul ignore next */
const serverLangPackCache = global.BROWSER ? null : require('./server-cache');

export function deferredForceLoad(locale, componentKey) {
  return {
    type: LANGUAGE_PACK_DEFERRED_FORCE_LOAD,
    locale,
    componentKey,
  };
}

const isLoaded = ({ getState, locale, componentKey }) => {
  const state = getState();
  return state.getIn(['intl', 'languagePacks', locale, componentKey])
    && !state.getIn(['intl', 'languagePacks', locale, componentKey, 'isLoading'])
    && !state.getIn(['intl', 'languagePacks', locale, componentKey, 'error']);
};

const getLoadingPromise = ({ getState, locale, componentKey }) => {
  const state = getState();
  return state.getIn(['intl', 'languagePacks', locale, componentKey, 'promise']);
};

const langPackToIguazu = ({ getState, componentKey }) => {
  const state = getState();
  const locale = state.getIn(['intl', 'activeLocale']);
  const langPackState = state.getIn(['intl', 'languagePacks', locale, componentKey], iMap({
    data: iMap(),
  }));
  return {
    locale,
    status: langPackState.get('isLoading', true) ? 'loading' : 'complete',
    data: langPackState.get('data').toJS(),
    error: langPackState.get('error'),
    errorExpiration: langPackState.get('errorExpiration'),
  };
};

const getUrl = ({ getState, langPackLocale, componentKey }) => {
  const state = getState();
  const moduleBaseUrl = getModuleBaseUrl(componentKey);
  const localeFilename = state.getIn(['config', 'localeFilename']);

  return `${[
    moduleBaseUrl,
    langPackLocale.toLowerCase(),
    localeFilename || componentKey,
  ].join('/')}.json`;
};

const fetchLanguagePack = ({
  dispatch,
  getState,
  url,
  locale,
  fallbackLocale,
  fallbackUrl,
  componentKey,
  fetchClient,
  retry = false,
}) => {
  let defaultUrl;
  if (!global.BROWSER) {
    defaultUrl = getUrl({ getState, langPackLocale: locale, componentKey });
    const cached = serverLangPackCache.get(url, defaultUrl);
    if (cached) {
      // eslint-disable-next-line no-console -- logging
      console.debug('using serverLangPackCache for %s', url);
      return Promise.resolve(cached);
    }
  }

  return fetchClient(url)
    .then((response) => {
      if (response.ok) {
        return response.json();
      }
      return Promise.reject(response);
    })
    .then((data) => {
      if (!global.BROWSER) {
        // eslint-disable-next-line no-console -- logging
        console.debug('setting serverLangPackCache: url %s, data %o', url, data);
        serverLangPackCache.set(url, data, defaultUrl);
      }
      return data;
    })
    .catch((errorOrResponse) => {
      if (errorOrResponse instanceof Error) {
        return Promise.reject(errorOrResponse);
      }

      const { status, statusText, url: responseUrl } = errorOrResponse;

      if (status === 404 && fallbackLocale && locale !== fallbackLocale && !retry) {
        // eslint-disable-next-line no-console -- logging
        console.warn('Missing %s language pack for %s, falling back to %s.', locale, componentKey, fallbackLocale);
        return fetchLanguagePack({
          getState,
          dispatch,
          url: fallbackUrl || getUrl({
            getState,
            langPackLocale: fallbackLocale,
            componentKey,
          }),
          locale: fallbackLocale,
          fetchClient,
          componentKey,
          retry,
        });
      }

      if (status === 404) return Promise.resolve({});

      const error = new Error(`${statusText} (${responseUrl})`);
      error.response = errorOrResponse;

      return Promise.reject(error);
    });
};

function deferredForcedLoadLanguagePack({
  dispatch,
  locale,
  url,
  fallbackLocale,
  fallbackUrl,
  componentKey,
  loadLanguagePackAction,
}) {
  if (!global.BROWSER) return;

  dispatch(deferredForceLoad(locale, componentKey));

  const callback = () => dispatch(loadLanguagePackAction(componentKey, {
    locale,
    url,
    force: true,
    fallbackLocale,
    fallbackUrl,
  }));

  if (typeof window.requestIdleCallback === 'function') {
    window.requestIdleCallback(callback);
  } else {
    setTimeout(callback, 10e6);
  }
}

const getResourceFromState = ({
  dispatch,
  getState,
  locale,
  url,
  fallbackLocale,
  fallbackUrl,
  componentKey,
  loadLanguagePackAction,
}) => {
  const state = getState();
  const langPack = state.getIn(['intl', 'languagePacks', locale, componentKey]);
  // eslint-disable-next-line no-underscore-dangle -- properties have underscore
  if (langPack && langPack._loadedOnServer && !langPack._pendingDeferredForceLoad) {
    // Make a request on the client once idle even though the language
    // pack was loaded on the server, so that it can be cached by the
    // service worker for offline browsing.
    deferredForcedLoadLanguagePack({
      dispatch,
      locale,
      url,
      fallbackLocale,
      fallbackUrl,
      componentKey,
      loadLanguagePackAction,
    });
  }
  return langPack;
};

export function loadLanguagePack(
  componentKey,
  {
    locale: givenLocale,
    force = false,
    url,
    fallbackLocale,
    fallbackUrl,
  } = {}
) {
  return (dispatch, getState, { fetchClient }) => {
    const intlState = getState().get('intl');
    const locale = givenLocale
      || intlState.get('nextLocale')
      || intlState.get('activeLocale');
    if (!locale) {
      return Promise.reject(new Error('Failed to load language pack. No locale was set or given'));
    }
    if (fallbackUrl && !fallbackLocale) {
      return Promise.reject(new Error('Fallback locale is required when fallback URL is provided for language pack'));
    }

    const existingPromise = getLoadingPromise({ getState, locale, componentKey });

    if (existingPromise && !force) {
      return existingPromise;
    }

    if (isLoaded({ getState, locale, componentKey }) && !force) {
      return Promise.resolve(getResourceFromState({
        dispatch,
        getState,
        locale,
        url,
        fallbackLocale,
        fallbackUrl,
        componentKey,
        loadLanguagePackAction: loadLanguagePack,
      }));
    }
    const promise = fetchLanguagePack({
      getState,
      dispatch,
      url: url || getUrl({
        getState,
        langPackLocale: locale,
        componentKey,
      }),
      locale,
      fallbackLocale,
      fallbackUrl,
      fetchClient,
      componentKey,
    });

    // Dispatch request
    dispatch({
      type: LANGUAGE_PACK_REQUEST,
      componentKey,
      locale,
      promise,
    });

    // Handle Success or Failure Dispatch as Side Effect
    promise
      .then((data) => dispatch({
        type: LANGUAGE_PACK_SUCCESS,
        lastFetched: Date.now(),
        componentKey,
        locale,
        data: fromJS(data),
      }))
      .catch((error) => dispatch({
        type: LANGUAGE_PACK_FAILURE,
        error,
        componentKey,
        locale,
      }))
      .finally(() => getResourceFromState({
        dispatch,
        getState,
        locale,
        url,
        fallbackLocale,
        fallbackUrl,
        componentKey,
        loadLanguagePackAction: loadLanguagePack,
      }));
    // Return original promise to have custom handling possible
    return promise;
  };
}

export function queryLanguagePack(componentKey, { fallbackLocale } = {}) {
  return (dispatch, getState) => {
    const iguazuState = langPackToIguazu({ getState, componentKey });
    if (iguazuState.error) {
      const isExpired = Date.now() > iguazuState.errorExpiration;
      if (!isExpired) {
        return {
          ...iguazuState,
          promise: Promise.reject(iguazuState.error),
        };
      }
    }
    const promise = dispatch(loadLanguagePack(componentKey, {
      locale: iguazuState.locale,
      fallbackLocale,
    }));
    return {
      ...langPackToIguazu({ getState, componentKey }),
      promise,
    };
  };
}

export const loadedLocales = new Set();

export function getLocalePack(locale) {
  const localeArray = locale.split('-');
  let localePack;

  while (localeArray.length > 0 && !localePack) {
    localePack = localePacks[localeArray.join('-')];
    localeArray.pop();
  }

  if (localePack) {
    if (!loadedLocales.has(locale)) {
      loadedLocales.add(locale);
      return localePack();
    }
    return Promise.resolve();
  }

  return Promise.reject(new Error(`No locale bundle available for "${locale}" (type ${typeof locale})`));
}

export function updateLocale(locale) {
  return (dispatch) => {
    if (!locale) {
      return Promise.reject(new Error('No locale was given'));
    }

    // supporting files for locale dependent libraries
    return getLocalePack(locale)
      .then(() => {
        dispatch({
          type: UPDATE_LOCALE,
          locale,
        });
      });
  };
}
