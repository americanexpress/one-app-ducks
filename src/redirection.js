/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { fromJS } from 'immutable';
import typeScope from './utils/typeScope';

export const SERVER_SIDE_REDIRECT = `${typeScope}/redirection/SERVER_SIDE_REDIRECT`;
export const CLIENT_SIDE_REDIRECT = `${typeScope}/redirection/CLIENT_SIDE_REDIRECT`;
export const NOOP = `${typeScope}/redirection/NOOP`;

export const initialState = fromJS({
  destination: null,
  redirectionInFlight: false,
});

export default function redirectionReducer(state = initialState, action) {
  switch (action.type) {
    case SERVER_SIDE_REDIRECT:
      return state.get('destination')
        ? state
        : state.set('destination', action.destination);
    case CLIENT_SIDE_REDIRECT:
      return state.set('redirectionInFlight', true);
    default:
      return state;
  }
}

export function externalRedirect(destination) {
  if (typeof destination !== 'string') {
    return { type: NOOP };
  }

  if (global.BROWSER) {
    window.location.assign(destination);
    return {
      type: CLIENT_SIDE_REDIRECT,
    };
  }

  return {
    type: SERVER_SIDE_REDIRECT,
    destination,
  };
}
