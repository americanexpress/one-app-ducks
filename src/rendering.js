/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { fromJS } from 'immutable';
import typeScope from './utils/typeScope';

export const SET_DANGEROUSLY_DISABLE_SCRIPTS = `${typeScope}/render/SET_DANGEROUSLY_DISABLE_SCRIPTS`;
export const SET_DANGEROUSLY_DISABLE_SCRIPTS_AND_STYLES = `${typeScope}/render/SET_DANGEROUSLY_DISABLE_SCRIPTS_AND_STYLES`;
export const SET_RENDER_PARTIAL_ONLY = `${typeScope}/render/SET_RENDER_PARTIAL_ONLY`;
export const SET_RENDER_TEXT_ONLY = `${typeScope}/render/SET_RENDER_TEXT_ONLY`;

export const initialState = fromJS({
  disableStyles: false,
  disableScripts: false,
  renderPartialOnly: false,
  renderTextOnly: false,
  renderTextOnlyOptions: { htmlTagReplacement: '', allowedHtmlTags: [] },
});
// eslint-disable-next-line default-param-last -- reducers have default params first
export default function reducer(state = initialState, action) {
  switch (action.type) {
    case SET_DANGEROUSLY_DISABLE_SCRIPTS:
      return state.set('disableScripts', action.disableScripts);
    case SET_DANGEROUSLY_DISABLE_SCRIPTS_AND_STYLES:
      return state
        .set('disableStyles', action.disableScriptsAndStyles)
        .set('disableScripts', action.disableScriptsAndStyles);
    case SET_RENDER_PARTIAL_ONLY:
      return state.set('renderPartialOnly', action.renderPartialOnly);
    case SET_RENDER_TEXT_ONLY:
      return state.withMutations((newState) => newState
        .set('renderTextOnly', action.renderTextOnly)
        .setIn(['renderTextOnlyOptions', 'htmlTagReplacement'], action.options.htmlTagReplacement)
        .setIn(['renderTextOnlyOptions', 'allowedHtmlTags'], action.options.allowedHtmlTags));
    default:
      return state;
  }
}

export const setDangerouslyDisableScripts = (disableScripts) => ({
  type: SET_DANGEROUSLY_DISABLE_SCRIPTS,
  disableScripts,
});

export const setDangerouslyDisableScriptsAndStyles = (disableScriptsAndStyles) => ({
  type: SET_DANGEROUSLY_DISABLE_SCRIPTS_AND_STYLES,
  disableScriptsAndStyles,
});

export const setRenderPartialOnly = (renderPartialOnly) => ({
  type: SET_RENDER_PARTIAL_ONLY,
  renderPartialOnly,
});

export const setRenderTextOnly = (renderTextOnly, options) => ({
  type: SET_RENDER_TEXT_ONLY,
  renderTextOnly,
  options: { htmlTagReplacement: '', allowedHtmlTags: [], ...options },
});
