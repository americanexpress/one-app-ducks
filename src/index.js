/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import browser, { setOrigin, setUserAgent, getCookies } from './browser';
import error, { applicationError, clearError } from './error';
import errorReporting, { addErrorToReport, sendErrorReport } from './errorReporting';
import intlReducer, {
  getLocalePack,
  loadLanguagePack,
  queryLanguagePack,
  updateLocale,
} from './intl';
import redirection, { externalRedirect } from './redirection';
import rendering, {
  setDangerouslyDisableScripts,
  setDangerouslyDisableScriptsAndStyles,
  setRenderPartialOnly,
  setRenderTextOnly,
} from './rendering';

export default {
  browser,
  error,
  errorReporting,
  intl: intlReducer,
  redirection,
  rendering,
};

export {
  // browser
  setOrigin,
  setUserAgent,
  getCookies,

  // error
  applicationError,
  clearError,

  // errorReporting
  addErrorToReport,
  sendErrorReport,

  // intl
  getLocalePack,
  loadLanguagePack,
  queryLanguagePack,
  updateLocale,

  // redirection
  externalRedirect,

  // rendering
  setDangerouslyDisableScripts,
  setDangerouslyDisableScriptsAndStyles,
  setRenderPartialOnly,
  setRenderTextOnly,
};
