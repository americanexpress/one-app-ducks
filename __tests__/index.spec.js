/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { createStore } from 'redux';
import { combineReducers } from 'redux-immutable';
import { fromJS } from 'immutable';

import * as publicAPI from '../src';

jest.mock('../src/intl/localePacks.client', () => require('../lib/intl/localePacks.client'), { virtual: true });
jest.mock('../src/intl/localePacks.node', () => require('../lib/intl/localePacks.node'), { virtual: true });

describe('index', () => {
  it('should expose the publicAPI', () => {
    expect(Object.keys(publicAPI)).toMatchSnapshot();
  });
});

describe('immutable store test', () => {
  const { default: reducers } = publicAPI;

  function isImmutableDeep(state) {
    return fromJS(state).equals(state);
  }

  it('should initialize the state for all reducers with immutable data types', () => {
    expect.assertions(2);

    const store = createStore(combineReducers(reducers));
    const state = store.getState();

    expect(isImmutableDeep(state)).toBe(true);
    expect(state).toMatchSnapshot();
  });

  it('should merge initial immutable state with missing initial state', () => {
    expect.assertions(2);

    const store = createStore(combineReducers(reducers), fromJS({}));
    const state = store.getState();

    expect(isImmutableDeep(state)).toBe(true);
    expect(state).toMatchSnapshot();
  });

  it('should fail to initialize if provided non-immutable initial state', () => {
    expect.assertions(1);

    expect(() => {
      createStore(combineReducers(reducers), {});
    }).toThrowErrorMatchingSnapshot();
  });
});
