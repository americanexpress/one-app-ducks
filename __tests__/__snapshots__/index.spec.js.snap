// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`immutable store test should fail to initialize if provided non-immutable initial state 1`] = `"inputState.withMutations is not a function"`;

exports[`immutable store test should initialize the state for all reducers with immutable data types 1`] = `
Immutable.Map {
  "browser": Immutable.Map {
    "cookies": null,
    "location": Immutable.Map {
      "origin": null,
      "host": null,
      "protocol": null,
    },
    "userAgent": null,
  },
  "error": Immutable.Map {},
  "errorReporting": Immutable.Map {
    "queue": Immutable.List [],
    "pending": Immutable.List [],
    "pendingPromise": null,
    "retryWait": 500,
  },
  "intl": Immutable.Map {
    "activeLocale": "en-US",
  },
  "redirection": Immutable.Map {
    "destination": null,
    "redirectionInFlight": false,
  },
  "rendering": Immutable.Map {
    "disableStyles": false,
    "disableScripts": false,
    "renderPartialOnly": false,
  },
}
`;

exports[`immutable store test should merge initial immutable state with missing initial state 1`] = `
Immutable.Map {
  "browser": Immutable.Map {
    "cookies": null,
    "location": Immutable.Map {
      "origin": null,
      "host": null,
      "protocol": null,
    },
    "userAgent": null,
  },
  "error": Immutable.Map {},
  "errorReporting": Immutable.Map {
    "queue": Immutable.List [],
    "pending": Immutable.List [],
    "pendingPromise": null,
    "retryWait": 500,
  },
  "intl": Immutable.Map {
    "activeLocale": "en-US",
  },
  "redirection": Immutable.Map {
    "destination": null,
    "redirectionInFlight": false,
  },
  "rendering": Immutable.Map {
    "disableStyles": false,
    "disableScripts": false,
    "renderPartialOnly": false,
  },
}
`;

exports[`index should expose the publicAPI 1`] = `
Array [
  "setOrigin",
  "setUserAgent",
  "getCookies",
  "applicationError",
  "clearError",
  "addErrorToReport",
  "sendErrorReport",
  "getLocalePack",
  "loadLanguagePack",
  "queryLanguagePack",
  "updateLocale",
  "externalRedirect",
  "setDangerouslyDisableScripts",
  "setDangerouslyDisableScriptsAndStyles",
  "setRenderPartialOnly",
  "default",
]
`;
